src/node/share_manager.h

```diff
@@ -329,25 +354,36 @@ namespace ccf
 
       LedgerSecretsMap restored_ledger_secrets;
 
-      auto restored_ls = ls_wrapping_key.unwrap(
-        recovery_shares_info->wrapped_latest_ledger_secret.encrypted_data);
+      auto restored_ls = combine_from_submitted_shares(tx).unwrap(
+        recovery_shares_info->wrapped_latest_ledger_secret);
       auto decryption_key = restored_ls.raw_key;
 
+      LOG_DEBUG_FMT(
+        "Recovering {} encrypted ledger secrets",
+        recovery_ledger_secrets.size());
+
+      auto& current_ledger_secret_version =
+        recovery_ledger_secrets.back().next_version;
+      if (!current_ledger_secret_version.has_value())
```

Comment:
If that should always be set, why is that optional? When is it ever unset?

Reply:
It is optional because the elements in `recovery_ledger_secrets` are of the same type as the values in the `encrypted_ledger_secrets` table (i.e. `PreviousLedgerSecretInfo`), for convenience mostly. The recovery hook on that table make sure that this value is set though. It's not set in the case of a rekey, when CCF doesn't know the value at which the new ledger secret is applicable when it writes the transaction, and the value is then set to the recovery hook version (see comment in the recovery hook in `node_state.h`)

Reply:
That seems unfortunate, perhaps we ought to consider splitting the type between the tables so we can be more precise?

Reply:
As discussed, I've added a comment to explain why this optional here is always set by the recovery hook. It is optional in the first place because the value for that table contains an optional.